CoordinatorTests.testDiscoveryUsesNodesFromLastClusterState test failure

Reproduces locally

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+multijob-unix-compatibility/os=debian-8/155/consoleFull

Failure:
```
13:54:46 org.elasticsearch.cluster.coordination.CoordinatorTests > testDiscoveryUsesNodesFromLastClusterState FAILED
13:54:46     java.lang.AssertionError: node1 has applied its state 
13:54:46     Expected: <606L>
13:54:46          but: was <605L>
13:54:46         at __randomizedtesting.SeedInfo.seed([567D9E1ADC657714:3060AB7CA14A7F0B]:0)
13:54:46         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
13:54:46         at org.junit.Assert.assertThat(Assert.java:956)
13:54:46         at org.elasticsearch.cluster.coordination.CoordinatorTests$Cluster.stabilise(CoordinatorTests.java:1524)
13:54:46         at org.elasticsearch.cluster.coordination.CoordinatorTests$Cluster.stabilise(CoordinatorTests.java:1505)
13:54:46         at org.elasticsearch.cluster.coordination.CoordinatorTests.testDiscoveryUsesNodesFromLastClusterState(CoordinatorTests.java:1086)
```

Reproduce:
```
 ./gradlew :server:test --tests "org.elasticsearch.cluster.coordination.CoordinatorTests.testDiscoveryUsesNodesFromLastClusterState" -Dtests.seed=567D9E1ADC657714 -Dtests.security.manager=true -Dtests.locale=es-US -Dtests.timezone=Chile/EasterIsland -Dcompiler.java=12 -Druntime.java=8
```

Seems to only fail on 7.x branch. Verified that `master` without the `java=8` flag it passed just fine.

--- a/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
+++ b/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
@@ -1231,7 +1231,7 @@ public class Coordinator extends AbstractLifecycleComponent implements Discovery
 
                 @Override
                 public String toString() {
-                    return "scheduled timeout for " + this;
+                    return "scheduled timeout for " + CoordinatorPublication.this;
                 }
             }, publishTimeout, Names.GENERIC);
         }

--- a/server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java
@@ -1515,6 +1515,10 @@ public class CoordinatorTests extends ESTestCase {
 
             final ClusterNode leader = getAnyLeader();
             final long leaderTerm = leader.coordinator.getCurrentTerm();
+
+            final int pendingTaskCount = leader.masterService.getFakeMasterServicePendingTaskCount();
+            runFor((pendingTaskCount + 1) * DEFAULT_CLUSTER_STATE_UPDATE_DELAY, "draining task queue");
+
             final Matcher<Long> isEqualToLeaderVersion = equalTo(leader.coordinator.getLastAcceptedState().getVersion());
             final String leaderId = leader.getId();
 
@@ -1527,6 +1531,8 @@ public class CoordinatorTests extends ESTestCase {
                 assertFalse(nodeId + " should not have an active publication", clusterNode.coordinator.publicationInProgress());
 
                 if (clusterNode == leader) {
+                    assertThat(nodeId + " is still the leader", clusterNode.coordinator.getMode(), is(LEADER));
+                    assertThat(nodeId + " did not change term", clusterNode.coordinator.getCurrentTerm(), is(leaderTerm));
                     continue;
                 }
 

--- a/server/src/test/java/org/elasticsearch/indices/cluster/FakeThreadPoolMasterService.java
+++ b/server/src/test/java/org/elasticsearch/indices/cluster/FakeThreadPoolMasterService.java
@@ -84,6 +84,10 @@ public class FakeThreadPoolMasterService extends MasterService {
         };
     }
 
+    public int getFakeMasterServicePendingTaskCount() {
+        return pendingTasks.size();
+    }
+
     private void scheduleNextTaskIfNecessary() {
         if (taskInProgress == false && pendingTasks.isEmpty() == false && scheduledNextTask == false) {
             scheduledNextTask = true;

